Setup Wizard: More helpful error state when trying to import from unprovisioned device

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->


https://github.com/learningequality/kolibri/assets/6356129/f3bde730-74f2-490d-8479-24170660706f


- Takes string [from here](https://github.com/learningequality/kolibri/blob/2692958f7772b099cf9191ff47e685ff4ca83206/kolibri/plugins/user_profile/assets/src/views/ChangeFacility/SelectFacility.vue#L294-L298) "No learning facilities found"
- Uses that by making the "Select facility" show an error when there are no facilities
- Adds `hideContinue` prop to `OnboardingStepBase` to hide the Continue button there when we have this error state
- Fixes issue where user's selection was lost when clicking "Go back" from the error state

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

Fixes #12066

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

See the issue comment here for [replication path](https://github.com/learningequality/kolibri/issues/12066#issuecomment-2104315832) -- note that I kinda had to go back and forth a few times through the flow - never actually importing anything - before I got the device to show up.

Overall, does the workflow feel okay?

nucleogenesis committed:
use borrowed string in recoverable error state when device has no facility

nucleogenesis committed:
initialize JoinOrNewLOD component w/ value from machine context (save user's selection between navigations)

pcenov commented:
Thanks @nucleogenesis - I confirm that now there's an error message if I attempt to import a learner from unprovisioned device.
Just have one question - in https://github.com/learningequality/kolibri/issues/12066 I have mentioned that the expected behavior would be for the device to not be shown in the 'Select device' modal which is the case for example when I attempt to import a full facility from that same unprovisioned device. Is that not an option in this case?

bjester commented:
To @pcenov's note, we should probably add a prop to `SelectDeviceForm` for filtering locations that have a facility, i.e. `filterHasFacility`. Then have the [`useDeviceFacilityFilter`](https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/useDevices.js#L178) apply the filter. That should prevent it from ever getting as far as where these error messages would show.

That would take care of filtering devices that would not be joinable or importable because they have no facility, which is a solid improvement. Although it doesn't go quite as far as the original issue's proposal of filtering those devices that actually have learners for import. I think the consensus was that we don't need to go that far. See my [comment](https://github.com/learningequality/kolibri/issues/12066#issuecomment-2072784354) in regards to that

nucleogenesis committed:
update typedef

nucleogenesis committed:
use device filtering to preemptively remove devices from list if they have no facilities by prop

nucleogenesis committed:
use i18n string for error message

nucleogenesis commented on kolibri/core/discovery/utils/network/client.py:
```diff
@@ -115,13 +115,13 @@ def post(self, path, **kwargs):
-        if "timeout" not in kwargs:
-            kwargs.update(timeout=self.timeout)
+
+        timeout = kwargs.pop("timeout", self.timeout)
```
@bjester I saw the comment in the filters about wanting to pass the params to the API directly and started investigating. I was running into an error when I tried to pass the filter values as kwargs and discovered that the parent request method expects the params kwarg.

Looking at changing the front-end code to pass these values in to the API seemed like maybe out of scope for this PR, but does this look right and like it might help simplify the facility filters a bit. I was thinking I might just put that into the `apiParams` object created in SelectDeviceForm but paused to get your thoughts here.

bjester replied:
Yeah so I think the changes that are needed are external to this client class. This class is a generic client utility. This means we rather need to update the APIs that are using it to pass the params.

nucleogenesis commented:
@radinamatic @bjester I've updated this with changes that I think are in line with Blaine's suggestion which hides no-facility devices altogether when a particular prop is passed.

I'm feeling a little unsure of my approach so I look forward to your thoughts on the code Blaine!

bjester commented on kolibri/core/discovery/utils/network/client.py:
```diff
@@ -115,13 +115,13 @@ def post(self, path, **kwargs):
         try:
             with super(NetworkClient, self).request(
-                method, url, stream=True, **kwargs
+                method, url, timeout=timeout, stream=True, params=kwargs
```
This is very likely a breaking change. Not all requests use `params`, and seeing as this overrides the base `request` method which the underlying methods call (`get`, `post`, etc), this change leads to it not properly passing all parameters to the super method.

nucleogenesis replied:
Ahh right on - I'll give this all a closer look and might ping you on Slack for a bit more clarifications around this - even if we revert this here it seems like some changes are needed around here in any case and I can make a follow-up issue

bjester commented on kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/SelectDeviceForm.vue:
```diff
@@ -166,9 +167,15 @@
         apiParams.subset_of_users_device = false;
+
+        if (props.filterByHasFacilities === true) {
+          deviceFilters.push(useDeviceHasFacilitiesFilter());
```
This works, although it does add some tech debt because if the prop was used in conjunction with the other facility filter props, we'd be calling the API twice. So ideally the existing `useDeviceFacilityFilter` is updated to have a special parameter supporting this type of filtering. Long term, the idea is to remove the filter props and use the filter composables directly in usages, which would likely pose the same problem.

nucleogenesis replied:
`useDeviceFacilityFilter` just seemed fundamentally different from what the new function is trying to do where it is basically saying "give me a facility-shaped thing and I'll find if there is a match" vs "are there _any_ facilities at all?" 

I considered updating the `useDeviceFacilityFilter` w/ a second param like `any=false` to mean like "oh actually we only care about if there are any facilities" but this felt stapling functionality onto something where it doesn't really belong.

I definitely get the issue w/ the tech debt here, though, and it _is_ a bit wonky as-is -- but I'm curious what you mean here:

> the idea is to remove the filter props and use the filter composables directly in usages

Like - would we instead be importing from `useDevices` in, say, the SetupWizard where we use the `<SelectDeviceForm>`, then we call the functions from `useDevices` to setup the filters and pass them directly, rather than working it out based on the props being null vs other?

bjester replied:
> fundamentally different from what the new function is trying to do where it is basically saying "give me a facility-shaped thing and I'll find if there is a match" vs "are there any facilities at all?"

I get that there's fundamentally different use cases, but IMO the logic isn't different enough. For example, I see it like this:

```js
// has matching facility with ID
let match = f => f.id = id;
facilities.filter(match).length > 0;

// has any
match = () => true;
facilities.filter(match).length > 0;
```

When a request to match a specific facility is applied, the 'any match' scenario becomes implicit, because having no facilities or a least a matching facility is the same 'any match' . So filtering by specific properties is really just a subset of 'any' match. The only important aspect would be to ensure that the logic represents a specific filter as a subset, meaning the specific filters take precedence over any match, which works because the results of that again are the same and do not conflict.

> Like - would we instead be importing from useDevices in, say, the SetupWizard where we use the <SelectDeviceForm>, then we call the functions from useDevices to setup the filters and pass them directly, rather than working it out based on the props being null vs other?

That's the idea-- to let the implementation use the composables and remove the complexity of handling and managing multiple props which could conflict or produce issues when used in conjunction.

bjester replied:
A compromise would be to keep it separate but just move where it's applied so it's impossible for both the specific filters and the any facility filter from being applied simultaneously. Which again works because of the same reasons

bjester commented on kolibri/plugins/setup_wizard/assets/src/views/SelectFacilityForm.vue:
```diff
@@ -99,6 +117,10 @@
+            if (this.facilities.length === 0) {
+              this.errorMessage = SelectFacilityStrings.$tr('noFacilitiesText');
+              return;
+            }
```
With the devices filtered in the modal, this should now be dead code (unreachable), right?

nucleogenesis replied:
_Should_ be yea :-D - I figure it's probably worth keeping in case something goes awry elsewhere. 

Maybe I should throw an error here - the upcoming distributed error reporting could come in handy here because we'd hear about it on Sentry (IIRC from the presentation this week).

bjester reviewed (changes requested):
More comments. The primary blocking change is the updates to the network client. I'm quite sure those changes are likely breaking.

nucleogenesis committed:
deviceHasMatchingFacility treats empty filter as wildcard

{} => "are there any facilities at all?"
{ id: <val> } => "are there any facilities that match this id?"

this ensures that we properly handle the case where there are no
facilities

bjester commented on kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/SelectDeviceForm.vue:
```diff
@@ -163,21 +164,25 @@
-        );
       }
 
+      deviceFilters.push(useDeviceFacilityFilter(facilityFilter));
```
I believe this should be:
```suggestion
      if (Object.keys(facilityFilter).length > 0 || props.filterByHasFacilities) {
        deviceFilters.push(useDeviceFacilityFilter(facilityFilter));
      }
```

nucleogenesis committed:
use filterByHasFacilities prop

bjester commented on kolibri/core/assets/src/views/sync/SelectDeviceModalGroup/SelectDeviceForm.vue:
```diff
@@ -163,19 +164,22 @@
+
+      // If we're filtering a particular facility
+      if (Object.keys(facilityFilter).length > 0 || props.filterByHasFacilities) {
         apiParams.subset_of_users_device = false;
```
It maybe unexpected that we filter out SoUDs when applying `props.filterByHasFacilities`, but since it's a new prop, we don't have any use cases, so it's def fine until we do.

bjester reviewed (approved):
Code LGTM

pcenov commented:
Hi @radinamatic - I confirm that the issue described in https://github.com/learningequality/kolibri/issues/12066 is fixed now and there are no new issues observed while regression testing.

radinamatic reviewed (approved):
Manual QA passes, good to go! 💯 👏🏽 :shipit:

rtibbles commented:
Readthedocs build appears to have stalled, and has nothing to do with this PR. Merging!
